src/testing/fast_array_utils/_array_type.py

```diff
@@ -289,14 +317,68 @@ def _to_cs_dataset(self, x: ArrayLike, /, *, dtype: DTypeLike | None = None) ->
         else:
             raise NotImplementedError
 
-        x_sparse = (
-            csr_array(np.asarray(x, dtype=dtype))
-            if self.cls is types.CSRDataset
-            else csc_array(np.asarray(x, dtype=dtype))
-        )
+        cls = csr_array if self.cls is types.CSRDataset else csc_array
+        x_sparse = self._to_scipy_sparse(x, dtype=dtype, cls=cls)
         anndata.io.write_elem(grp, "/", x_sparse)
         return anndata.io.sparse_dataset(grp)
 
+    def _to_scipy_sparse(
+        self,
+        x: ArrayLike | Array | spmatrix | types.CupySpMatrix,
+        /,
+        *,
+        dtype: DTypeLike | None = None,
+        cls: type[CSBase] | None = None,
+    ) -> types.CSBase:
+        """Convert to a scipy sparse matrix/array."""
+        from scipy.sparse import spmatrix
+
+        from fast_array_utils import types
+        from fast_array_utils.conv import to_dense
+
+        if isinstance(x, types.DaskArray):
+            x = x.compute()
+        if isinstance(x, types.CupySpMatrix):
+            x = x.get()  # can be a coo_matrix due to dask concatenation
+        elif not isinstance(x, spmatrix | np.ndarray):
+            x = to_dense(x, to_memory=True)
+
+        cls = cast("type[types.CSBase]", cls or self.cls)
+        return cls(x, dtype=dtype)  # type: ignore[arg-type]
+
+    def _to_cupy_array(
+        self, x: ArrayLike | Array, /, *, dtype: DTypeLike | None = None
+    ) -> types.CupyArray:
+        import cupy as cu
+
+        from fast_array_utils import types
+        from fast_array_utils.conv import to_dense
+
+        if isinstance(x, types.DaskArray):
+            x = x.compute()  # this could now be a cupy array already
+        if isinstance(x, types.CupySpMatrix):
+            x = x.toarray()
+        if isinstance(x, types.CSDataset | types.CSBase):
+            x = to_dense(x, to_memory=True)
+
+        return cu.asarray(x, dtype=None if dtype is None else np.dtype(dtype))
+
+    def _to_cupy_sparse(
+        self,
+        x: ArrayLike | Array | spmatrix | types.CupySpMatrix,
+        /,
+        *,
+        dtype: DTypeLike | None = None,
+    ) -> types.CupyCSMatrix:
+        from scipy.sparse import spmatrix
+
+        from fast_array_utils import types
```

Comment:
why import here?

Reply:
(and elsewhere these internal imports - circular?)

Reply:
No, lazy. I don’t want to import the actual package from the testing utils at module execution time.